### PR TITLE
[SPARK-48416][SQL] Support related nested WITH expression

### PR DESCRIPTION
**Pull request by zml1206**
### What changes were proposed in this pull request?
Expression definitions should apply to the entire `With` child contains internally nested `WITH`, so refactor `RewriteWithExpression` logic.
Rewrite `With` consists of two steps:
1. Convert defs into `Project` of inputPlans
2. Replace the references in inputPlans and child then output the new child

`With` support nestd, defs and child can all contain sub-With, the `Alias` generated by
the current `With` acts on the child and the child's internal `With`, so generate `Project` order is：
1. internally nested with of main expression definitions
2. main expression
4. internally nested with of main expression

### Why are the changes needed?
Optimize `WITH` nested to apply to more scenarios, such as push down predicate.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit test.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by cloud-fan** [comment]
Can we solve this issue by recursion? Which is easier to understand.

**Comment by zml1206** [comment]
> Can we solve this issue by recursion? Which is easier to understand.

I have optimized the code and comments, please take a look again.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala [hunk 1]
```diff
@@ -177,4 +120,83 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
...
+    val childProjections = Array.fill(inputPlans.length)(mutable.ArrayBuffer.empty[Alias])
+    defs.zipWithIndex.foreach { case (CommonExpressionDef(child, id), index) =>
+      if (id.canonicalized) {
+        throw SparkException.internalError(
+          "Cannot rewrite canonicalized Common expression definitions")
+      }
+
+      if (CollapseProject.isCheap(child)) {
+        refToExpr(id) = child
+      } else {
```
Comment (holdenk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:137]: This seems potentially less than ideal, it looks like CollapseProject.isCheap will return true for Python UDFS. It also _might_ violate the promise of being gauranteed to be evaluated only once (see With.scala L26).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala [hunk 2]
```diff
@@ -177,4 +120,83 @@ object RewriteWithExpression extends Rule[LogicalPlan] {
...
+
+  private def genProject(
+      defs: Seq[Expression],
+      inputPlans: Array[LogicalPlan]): mutable.Map[CommonExpressionId, Expression] = {
+    val refToExpr = mutable.HashMap.empty[CommonExpressionId, Expression]
+    val childProjections = Array.fill(inputPlans.length)(mutable.ArrayBuffer.empty[Alias])
+    defs.zipWithIndex.foreach { case (CommonExpressionDef(child, id), index) =>
+      if (id.canonicalized) {
+        throw SparkException.internalError(
+          "Cannot rewrite canonicalized Common expression definitions")
```
Comment (holdenk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteWithExpression.scala:132]: Can we include the element that caused this error?

**Review by holdenk — COMMENTED** [review]
I'm not very familiar with this chunk of code but had a quick question while looking at it.

**Comment by holdenk** [comment]
CC @cloud-fan

**Comment by cloud-fan** [comment]
Sorry for the late review. I think the implementation can be much simpler, please see my attempt at https://github.com/apache/spark/pull/49093

**Pull request closed without merging**